src/main/java/teammates/ui/webapi/action/GetCoursesAction.java

```diff
@@ -17,17 +24,73 @@ protected AuthType getMinAuthLevel() {
 
     @Override
     public void checkSpecificAccessControl() {
-        gateKeeper.verifyInstructorPrivileges(logic.getAccount(userInfo.getId()));
+        String entityType = getNonNullRequestParamValue(Const.ParamsNames.ENTITY_TYPE);
+
+        if (!((entityType.equals(Const.EntityType.STUDENT) && userInfo.isStudent)
+                || (entityType.equals(Const.EntityType.INSTRUCTOR) && userInfo.isInstructor))) {
+            throw new UnauthorizedAccessException("Current account cannot access to courses of request entity type");
+        }
     }
 
     @Override
     public ActionResult execute() {
-        List<CourseAttributes> courses =
-                logic.getCoursesForInstructor(logic.getInstructorsForGoogleId(userInfo.getId(), true));
+        String entityType = getNonNullRequestParamValue(Const.ParamsNames.ENTITY_TYPE);
+        switch (entityType) {
+        case Const.EntityType.STUDENT:
+            return getStudentCourses();
+        case Const.EntityType.INSTRUCTOR:
+            return getInstructorCourses();
+        default:
+            return new JsonResult("Error: invalid entity type", HttpStatus.SC_BAD_REQUEST);
+
+        }
+    }
+
+    private ActionResult getStudentCourses() {
+        List<CourseAttributes> courses = logic.getCoursesForStudentAccount(userInfo.id);
+        return new JsonResult(new CoursesData(courses));
```

Comment:
No params? I'm noting that GetStudentCoursesAction takes in a optional Const.ParamsNames.CHECK_PERSISTENCE_COURSE, which relates to how the server might handle if a student joins a course too early before this data has propogated fully to all the servers. I feel like if this is not handled it would be a regression from current functionality?

Reply:
I asked Xiao Pu about this. The reasons are:
1. Such eventual consistency should be handled by frontend. API does not handle this
2. Database is about to upgrade. `Firestore` will provide strong consistency level, so that this problem won't  happen.